share/man/man4/smb.4
.\" Copyright (c) 1998, Nicolas Souchu
.\" Copyright (c) 2004, Joerg Wunsch
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: head/share/man/man4/smb.4 213573 2010-10-08 12:40:16Z uqs $
.\"
.Dd February 6, 2009
.Dt SMB 4
.Os
.Sh NAME
.Nm smb
.Nd SMB generic I/O device driver
.Sh SYNOPSIS
.Cd "device smb"
.Sh DESCRIPTION
The
.Em smb
character device driver provides generic i/o to any
.Xr smbus 4
instance.
In order to control SMB devices, use
.Pa /dev/smb?
with the ioctls described below.
Any of these ioctl commands takes a pointer to
.Vt struct smbcmd
as its argument.
.Bd -literal
#include <sys/types.h>

struct smbcmd {
	char cmd;
	int count;
	u_char slave;
	union {
		char byte;
		short word;

		char *byte_ptr;
		short *word_ptr;

		struct {
			short sdata;
			short *rdata;
		} process;
	} data;
};
.Ed
.Pp
The
.Fa slave
field is always used, and provides the address of the
SMBus slave device to talk to.
The slave address is specified in the seven most significant bits
.Pq i.e. Dq "left-justified" .
The least significant bit of the slave address must be zero.
.Pp
.Bl -column ".Dv SMB_QUICK_WRITE" -compact
.It Em Ioctl Ta Em Description
.Pp
.It Dv SMB_QUICK_WRITE Ta
The
.Em QuickWrite
command just issues the device address with write intent
to the bus, without transferring any data.
.It Dv SMB_QUICK_READ Ta
The
.Em QuickRead
command just issues the device address with read intent
to the bus, without transferring any data.
.It Dv SMB_SENDB Ta
The
.Em SendByte
command sends the byte provided in the
.Fa cmd
field to the device.
.It Dv SMB_RECVB Ta
The
.Em ReceiveByte
command reads a single byte from the device which will
be returned in the
.Fa cmd
field.
.It Dv SMB_WRITEB Ta
The
.Em WriteByte
command first sends the byte from the
.Fa cmd
field to the device, followed by the byte given in
.Fa data.byte .
.It Dv SMB_WRITEW Ta
The
.Em WriteWord
command first sends the byte from the
.Fa cmd
field to the device, followed by the word given in
.Fa data.word .
Note that the SMBus byte-order is little-endian by definition.
.It Dv SMB_READB Ta
The
.Em ReadByte
command first sends the byte from the
.Fa cmd
field to the device, and then reads one byte of data from
the device.
The returned data will be stored in the location pointed to by
.Fa data.byte_ptr .
.It Dv SMB_READW Ta
The
.Em ReadWord
command first sends the byte from the
.Fa cmd
field to the device, and then reads one word of data from
the device.
The returned data will be stored in the location pointed to by
.Fa data.word_ptr .
.It Dv SMB_PCALL Ta
The
.Em ProcedureCall
command first sends the byte from the
.Fa cmd
field to the device, followed by the word provided in
.Fa data.process.sdata .
It then reads one word of data from the device, and returns it
in the location pointed to by
.Fa data.process.rdata .
.It Dv SMB_BWRITE Ta
The
.Em BlockWrite
command first sends the byte from the
.Fa cmd
field to the device, followed by
.Fa count
bytes of data that are taken from the buffer pointed to by
.Fa data.byte_ptr .
The SMBus specification mandates that no more than 32 bytes of
data can be transferred in a single block read or write command.
This value is available in the constant
.Dv SMB_MAXBLOCKSIZE .
.It Dv SMB_BREAD Ta
The
.Em BlockRead
command first sends the byte from the
.Fa cmd
field to the device, and then reads
.Fa count
bytes of data that from the device.
These data will be returned in the buffer pointed to by
.Fa data.byte_ptr .
.El
.Pp
The
.Xr read 2
and
.Xr write 2
system calls are not implemented by this driver.
.Sh ERRORS
The
.Xr ioctl 2
commands can cause the following driver-specific errors:
.Bl -tag -width Er
.It Bq Er ENXIO
Device did not respond to selection.
.It Bq Er EBUSY
Device still in use.
.It Bq Er ENODEV
Operation not supported by device (not supposed to happen).
.It Bq Er EINVAL
General argument error.
.It Bq Er EWOULDBLOCK
SMBus transaction timed out.
.El
.Sh SEE ALSO
.Xr ioctl 2 ,
.Xr smbus 4
.Sh HISTORY
The
.Nm
manual page first appeared in
.Fx 3.0 .
.Sh AUTHORS
This
manual page was written by
.An Nicolas Souchu .


sys/dev/smbus/smb.c
/*-
 * Copyright (c) 1998, 2001 Nicolas Souchu
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $FreeBSD: head/sys/dev/smbus/smb.c 188077 2009-02-03 16:14:37Z jhb $
 */

#ifdef HAVE_KERNEL_OPTION_HEADERS
#include "opt_compat.h"
#endif

#include <sys/param.h>
#include <sys/kernel.h>
#include <sys/systm.h>
#include <sys/module.h>
#include <sys/bus.h>
#include <sys/conf.h>
#include <sys/uio.h>
#include <sys/fcntl.h>

#include <dev/smbus/smbconf.h>
#include <dev/smbus/smbus.h>
#include <dev/smbus/smb.h>

#include "smbus_if.h"

#define BUFSIZE 1024

struct smb_softc {
	device_t sc_dev;
	int sc_count;			/* >0 if device opened */
	struct cdev *sc_devnode;
	struct mtx sc_lock;
};

static void smb_identify(driver_t *driver, device_t parent);
static int smb_probe(device_t);
static int smb_attach(device_t);
static int smb_detach(device_t);

static devclass_t smb_devclass;

static device_method_t smb_methods[] = {
	/* device interface */
	DEVMETHOD(device_identify,	smb_identify),
	DEVMETHOD(device_probe,		smb_probe),
	DEVMETHOD(device_attach,	smb_attach),
	DEVMETHOD(device_detach,	smb_detach),

	/* smbus interface */
	DEVMETHOD(smbus_intr,		smbus_generic_intr),

	{ 0, 0 }
};

static driver_t smb_driver = {
	"smb",
	smb_methods,
	sizeof(struct smb_softc),
};

static	d_open_t	smbopen;
static	d_close_t	smbclose;
static	d_ioctl_t	smbioctl;

static struct cdevsw smb_cdevsw = {
	.d_version =	D_VERSION,
	.d_flags =	D_TRACKCLOSE,
	.d_open =	smbopen,
	.d_close =	smbclose,
	.d_ioctl =	smbioctl,
	.d_name =	"smb",
};

static void
smb_identify(driver_t *driver, device_t parent)
{

	if (device_find_child(parent, "smb", -1) == NULL)
		BUS_ADD_CHILD(parent, 0, "smb", -1);
}

static int
smb_probe(device_t dev)
{
	device_set_desc(dev, "SMBus generic I/O");

	return (0);
}
	
static int
smb_attach(device_t dev)
{
	struct smb_softc *sc = device_get_softc(dev);

	sc->sc_dev = dev;
	sc->sc_devnode = make_dev(&smb_cdevsw, device_get_unit(dev),
	    UID_ROOT, GID_WHEEL, 0600, "smb%d", device_get_unit(dev));
	sc->sc_devnode->si_drv1 = sc;
	mtx_init(&sc->sc_lock, device_get_nameunit(dev), NULL, MTX_DEF);

	return (0);
}

static int
smb_detach(device_t dev)
{
	struct smb_softc *sc = (struct smb_softc *)device_get_softc(dev);

	if (sc->sc_devnode)
		destroy_dev(sc->sc_devnode);
	mtx_destroy(&sc->sc_lock);

	return (0);
}

static int
smbopen(struct cdev *dev, int flags, int fmt, struct thread *td)
{
	struct smb_softc *sc = dev->si_drv1;

	mtx_lock(&sc->sc_lock);
	if (sc->sc_count != 0) {
		mtx_unlock(&sc->sc_lock);
		return (EBUSY);
	}

	sc->sc_count++;
	mtx_unlock(&sc->sc_lock);

	return (0);
}

static int
smbclose(struct cdev *dev, int flags, int fmt, struct thread *td)
{
	struct smb_softc *sc = dev->si_drv1;

	mtx_lock(&sc->sc_lock);
	KASSERT(sc->sc_count == 1, ("device not busy"));
	sc->sc_count--;
	mtx_unlock(&sc->sc_lock);

	return (0);
}

static int
smbioctl(struct cdev *dev, u_long cmd, caddr_t data, int flags, struct thread *td)
{
	char buf[SMB_MAXBLOCKSIZE];
	device_t parent;
	struct smbcmd *s = (struct smbcmd *)data;
	struct smb_softc *sc = dev->si_drv1;
	device_t smbdev = sc->sc_dev;
	int error;
	short w;
	u_char count;
	char c;

	parent = device_get_parent(smbdev);

	/* Make sure that LSB bit is cleared. */
	if (s->slave & 0x1)
		return (EINVAL);

	/* Allocate the bus. */
	if ((error = smbus_request_bus(parent, smbdev,
			(flags & O_NONBLOCK) ? SMB_DONTWAIT : (SMB_WAIT | SMB_INTR))))
		return (error);

	switch (cmd) {
	case SMB_QUICK_WRITE:
		error = smbus_error(smbus_quick(parent, s->slave, SMB_QWRITE));
		break;

	case SMB_QUICK_READ:
		error = smbus_error(smbus_quick(parent, s->slave, SMB_QREAD));
		break;

	case SMB_SENDB:
		error = smbus_error(smbus_sendb(parent, s->slave, s->cmd));
		break;

	case SMB_RECVB:
		error = smbus_error(smbus_recvb(parent, s->slave, &s->cmd));
		break;

	case SMB_WRITEB:
		error = smbus_error(smbus_writeb(parent, s->slave, s->cmd,
						s->data.byte));
		break;

	case SMB_WRITEW:
		error = smbus_error(smbus_writew(parent, s->slave,
						s->cmd, s->data.word));
		break;

	case SMB_READB:
		if (s->data.byte_ptr) {
			error = smbus_error(smbus_readb(parent, s->slave,
						s->cmd, &c));
			if (error)
				break;
			error = copyout(&c, s->data.byte_ptr,
					sizeof(*(s->data.byte_ptr)));
		}
		break;

	case SMB_READW:
		if (s->data.word_ptr) {
			error = smbus_error(smbus_readw(parent, s->slave,
						s->cmd, &w));
			if (error == 0) {
				error = copyout(&w, s->data.word_ptr,
						sizeof(*(s->data.word_ptr)));
			}
		}
		break;

	case SMB_PCALL:
		if (s->data.process.rdata) {

			error = smbus_error(smbus_pcall(parent, s->slave, s->cmd,
				s->data.process.sdata, &w));
			if (error)
				break;
			error = copyout(&w, s->data.process.rdata,
					sizeof(*(s->data.process.rdata)));
		}
		
		break;

	case SMB_BWRITE:
		if (s->count && s->data.byte_ptr) {
			if (s->count > SMB_MAXBLOCKSIZE)
				s->count = SMB_MAXBLOCKSIZE;
			error = copyin(s->data.byte_ptr, buf, s->count);
			if (error)
				break;
			error = smbus_error(smbus_bwrite(parent, s->slave,
						s->cmd, s->count, buf));
		}
		break;

#if defined(COMPAT_FREEBSD4) || defined(COMPAT_FREEBSD5) || defined(COMPAT_FREEBSD6)
	case SMB_OLD_BREAD:
#endif
	case SMB_BREAD:
		if (s->count && s->data.byte_ptr) {
			count = min(s->count, SMB_MAXBLOCKSIZE);
			error = smbus_error(smbus_bread(parent, s->slave,
						s->cmd, &count, buf));
			if (error)
				break;
			error = copyout(buf, s->data.byte_ptr,
			    min(count, s->count));
			s->count = count;
		}
		break;
		
	default:
		error = ENOTTY;
	}

	smbus_release_bus(parent, smbdev);

	return (error);
}

DRIVER_MODULE(smb, smbus, smb_driver, smb_devclass, 0, 0);
MODULE_DEPEND(smb, smbus, SMBUS_MINVER, SMBUS_PREFVER, SMBUS_MAXVER);
MODULE_VERSION(smb, 1);
